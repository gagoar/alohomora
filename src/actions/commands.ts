export { command as listCommand } from './listParameters';
export { command as getCommand } from './getParameter';
export { command as setCommand } from './setParameter';
export { command as exportCommand } from './exportAsTemplate';
export { command as deleteCommand } from './deleteParameter';