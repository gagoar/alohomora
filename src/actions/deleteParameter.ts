import SSM from 'aws-sdk/clients/ssm';
import { AWSError } from 'aws-sdk/lib/core';
import ora from 'ora';

import { Actions } from '../types';
import { REGION, API_VERSION, Environment, SUCCESS_SYMBOL } from '../utils/constants';
import { getGlobalOptions, Command } from '../utils/getGlobalOptions';
import { setAWSCredentials } from '../utils/setAWSCredentials';


interface Input extends Actions {
  name: string;

};

export const deleteParameter = async ({ name, prefix, region = REGION, environment = Environment.all }: Input): Promise<void> => {

  const loader = ora(`deleting key ${name} with prefix /${prefix} (${region})`).start();

  const ssm = new SSM({ apiVersion: API_VERSION, region });

  const keyName = `/${prefix}/${environment}/${name}`;

  const params = {
    Name: keyName,
  }

  try {
    await ssm.deleteParameter(params).promise();
    loader.stopAndPersist({ text: `deleted ${keyName} (${region})`, symbol: SUCCESS_SYMBOL });
  } catch (e) {
    if ((e as AWSError).code === 'ParameterNotFound') {
      loader.stopAndPersist({ text: `parameter ${keyName} not found (${region})`, symbol: SUCCESS_SYMBOL });
    } else {
      loader.fail(`Something went wrong deleting the key ${keyName}, ${e}`);
    }
  }
}


export const command = async (name: string, command: Command): Promise<void> => {

  const { params, credentials } = await getGlobalOptions(command);

  setAWSCredentials(credentials);
  await deleteParameter({ ...params, name });
}