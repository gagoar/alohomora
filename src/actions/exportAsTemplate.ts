import SSM from 'aws-sdk/clients/ssm';
import groupBy from 'lodash.groupby';

import { Actions } from '../types';
import { REGION, API_VERSION, MAX_RESULTS_FOR_PATH, Environment, Template } from '../utils/constants';
import { normalizeSecretKey } from '../utils/normalizeSecrets';
import { isValidTemplate } from '../utils/guards';
import { getGlobalOptions, Command } from '../utils/getGlobalOptions';
import { setAWSCredentials } from '../utils/setAWSCredentials';
interface Input extends Actions { templateName?: Template, custom?: string };

const getParametersByPath = async (params: SSM.GetParametersByPathRequest, region: string): Promise<SSM.ParameterList> => {

  const ssm = new SSM({ apiVersion: API_VERSION, region });
  const { Parameters: parameters = [], NextToken } = await ssm.getParametersByPath(params).promise();

  if (NextToken) {
    const moreParameters = await getParametersByPath({ ...params, NextToken }, region);
    return [...parameters, ...moreParameters];
  } else {
    return parameters;
  }
}

const getKeys = async ({ prefix, region }: { prefix: string, region: string }) => {
  const parameters = await getParametersByPath({
    Path: `/${prefix}`,
    MaxResults: MAX_RESULTS_FOR_PATH,
    Recursive: true,
    WithDecryption: true,
  }, region);

  const secrets = parameters.map(({ Name = '', Value: value = '' }) => {

    const { name, environment } = normalizeSecretKey(Name, prefix);
    return {
      name,
      environment,
      value,
    }
  });

  return groupBy(secrets, ({ environment }: { environment: string }) => environment);
}
const templateFunctions = {
  [Template.shell]: (secrets: Record<string, string>): string => {
    return Object.keys(secrets).map(key => `export ${key}='${secrets[key]}'`).join('\n');
  },
  [Template.json]: (secrets: Record<string, string>): string => {
    return JSON.stringify(secrets, null, 2);
  },
}

type Keys = { name: string; environment: string; value: string; };

const normalizeGroup = (keys: Keys[]): Record<string, string> => {
  return keys.reduce(((memo, secret) => {
    return Object.keys(memo).length ? { ...memo, [secret.name]: secret.value } : { [secret.name]: secret.value };
  }), {} as Record<string, string>);
};

export const exportAsTemplate = async ({ prefix, environment = Environment.all, region = REGION, templateName = Template.shell }: Input): Promise<string> => {

  try {
    const groupedSecrets = await getKeys({
      prefix,
      region,
    });

    let secrets: Record<string, string>;
    const baseSecrets = normalizeGroup(groupedSecrets[Environment.all] || []);

    if (environment !== Environment.all && environment in groupedSecrets) {
      const envSecrets = normalizeGroup(groupedSecrets[environment]);

      secrets = { ...baseSecrets, ...envSecrets };
    } else {
      secrets = baseSecrets;
    }

    return templateFunctions[templateName](secrets);

  } catch (e) {
    console.error('We found an error trying to retrieve secrets', e);
    throw e;
  }
};

export const command = async (templateName: string = Template.shell, command: Command): Promise<void> => {

  if (isValidTemplate(templateName)) {
    const { params, credentials } = await getGlobalOptions(command);

    setAWSCredentials(credentials);
    const response = await exportAsTemplate({ ...params, templateName });
    console.log(response);
  } else {
    console.error(`please provide a valid templateName (${Template})`);
    process.exit(1)
  }

}
