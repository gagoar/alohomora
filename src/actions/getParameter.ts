import SSM from 'aws-sdk/clients/ssm';
import { AWSError } from 'aws-sdk/lib/core';
import dateFormat from 'dateformat'
import ora from 'ora';

import { Actions } from '../types';
import { REGION, API_VERSION, Environment, DATE_FORMAT, SUCCESS_SYMBOL, DISABLE_TABLE_COLORS } from '../utils/constants';
import { getGlobalOptions, Command } from '../utils/getGlobalOptions';
import { setAWSCredentials } from '../utils/setAWSCredentials';
import { createTable, getTableHeader } from '../utils/tables';


interface Input extends Actions {
  name: string
};

export const getParameter = async ({ name, prefix, region = REGION, ci = false, environment = Environment.all }: Input): Promise<string> => {

  const loader = ora(`retrieving key ${name} with the prefix /${prefix}  (${region})`).start();

  const ssm = new SSM({ apiVersion: API_VERSION, region });

  const params = {
    Name: `/${prefix}/${environment}/${name}`,
    WithDecryption: true
  };

  let response: SSM.GetParameterResult;

  try {
    response = await ssm.getParameter(params).promise();

    if (response.Parameter) {

      const { Value: value, LastModifiedDate: updatedAt, Version: version } = response.Parameter;

      const style = ci ? DISABLE_TABLE_COLORS : undefined;
      const content = createTable(getTableHeader(), [[name, value, environment, dateFormat(updatedAt, DATE_FORMAT), version]], style)

      loader.stopAndPersist({ text: `${name} found under /${prefix}  (${region})`, symbol: SUCCESS_SYMBOL });

      return content.toString();

    } else {
      loader.stopAndPersist({ text: `${name} not found under /${prefix}  (${region})`, symbol: SUCCESS_SYMBOL });
    }

  } catch (e) {
    if ((e as AWSError).code === 'ParameterNotFound') {
      loader.stopAndPersist({ text: `${name} not found under /${prefix}  (${region})`, symbol: SUCCESS_SYMBOL });
    } else {
      loader.fail(`something went wrong retrieving the key ${name}, ${e}`);
    }
  }

  return '';
}

export const command = async (name: string, command: Command): Promise<void> => {

  const { params, credentials } = await getGlobalOptions(command);

  setAWSCredentials(credentials);
  const response = await getParameter({ ...params, name });
  console.log(response);
}