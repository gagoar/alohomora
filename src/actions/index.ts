export { listParameters } from './listParameters';
export { getParameter } from './getParameter';
export { setParameter } from './setParameter';
export { deleteParameter } from './deleteParameter';
export { exportAsTemplate } from './exportAsTemplate';