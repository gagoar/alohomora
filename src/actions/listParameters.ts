import Table, { HorizontalAlignment } from 'cli-table3';
import colors from 'colors/safe';
import dateFormat from 'dateformat';
import ora from 'ora';
import SSM from 'aws-sdk/clients/ssm';

import { Actions } from '../types';
import { API_VERSION, REGION, DATE_FORMAT, SUCCESS_SYMBOL, MAX_RESULTS_FOR_DESCRIBE, DISABLE_TABLE_COLORS } from '../utils/constants';
import { normalizeSecretKey } from '../utils/normalizeSecretKey';
import { Command, getGlobalOptions } from '../utils/getGlobalOptions';
import { setAWSCredentials } from '../utils/setAWSCredentials';
import group from 'lodash.groupby';

const getUser = (lastModifiedUser: string): string => {
  return lastModifiedUser.split('user/')[1];
}

const describeParameters = async (params: SSM.DescribeParametersRequest, region: string): Promise<SSM.ParameterMetadataList> => {
  const ssm = new SSM({ apiVersion: API_VERSION, region });
  const { Parameters: parameters = [], NextToken } = await ssm.describeParameters(params).promise();

  if (NextToken) {
    const moreParameters = await describeParameters({ ...params, NextToken }, region);
    return [...parameters, ...moreParameters];
  } else {
    return parameters;
  }
}

enum GroupBy {
  name = 'name',
  environment = 'environment'
}
interface Input extends Actions {
  groupBy?: keyof typeof GroupBy
}


const HEADERS: Record<keyof typeof GroupBy, string[]> = {
  'environment': ['Name', 'Updated by', 'Updated at'],
  'name': ['Environment', 'Updated by', 'Updated at']
}

const getTableHeader = (groupBy: Input['groupBy']) => {
  if (!groupBy) {
    return ['Name', 'Environment', 'Updated by', 'Updated at'];
  }

  return HEADERS[groupBy];
}

type TableContent = Table.HorizontalTableRow | Table.VerticalTableRow | Table.CrossTableRow;
const createTable = (head: Table.TableOptions['head'] | undefined, content: TableContent[], style?: Table.TableConstructorOptions['style']) => {

  const table = new Table({
    head,
    style
  });

  table.push(...content);
  return table;
}

export const listParameters = async ({ environment, prefix, region = REGION, ci = false, groupBy }: Input): Promise<string> => {


  const content = [];
  const loader = ora(`Finding keys with the prefix /${prefix}  (${region})`).start();

  const path = environment ? `/${prefix}/${environment}/` : `/${prefix}`;

  const parameterFilters = [
    {
      Key: 'Name',
      Option: 'Contains',
      Values: [path]
    }
  ]

  const params = {
    MaxResults: MAX_RESULTS_FOR_DESCRIBE,
    ParameterFilters: parameterFilters,
  };


  let parameters: SSM.ParameterMetadataList = [];

  try {
    parameters = await describeParameters(params, region);
    const keys = parameters.map(({ Name = '', LastModifiedDate = '', LastModifiedUser = '' }) => {
      const { name, environment } = normalizeSecretKey(Name, prefix);
      const date = dateFormat(LastModifiedDate, DATE_FORMAT);
      return [name, environment, getUser(LastModifiedUser), date];
    });

    keys.sort((a, b) => {
      if (a[0] > b[0]) return 1;
      if (a[0] < b[0]) return -1;
      return 0;
    })

    const styles = ci ? DISABLE_TABLE_COLORS : undefined;

    if (!groupBy) {
      const table = createTable(getTableHeader(groupBy), keys, styles);
      content.push(table.toString());
    } else {
      const groups = group(keys, tuple => groupBy === GroupBy.name ? tuple[0] : tuple[1]);

      const finalGroups = Object.keys(groups).map(group => {
        const values = groups[group].map(tuple =>
          groupBy === GroupBy.name ? [tuple[1], tuple[2], tuple[3]] : [tuple[0], tuple[2], tuple[3]]
        );

        const label = `${GroupBy[groupBy]}: ${group}`;

        const groupByLabel = [{ colSpan: 3, hAlign: 'center' as HorizontalAlignment, content: ci ? label : colors.red(label) }];
        const header = ci ? getTableHeader(groupBy) : getTableHeader(groupBy).map(value => colors.red(value));

        const table = createTable(undefined, [groupByLabel, header, ...values], styles);

        return table.toString();
      });
      content.push(...finalGroups);
    }

    loader.stopAndPersist({ text: `found ${parameters.length} secrets, under /${prefix}  (${region})`, symbol: SUCCESS_SYMBOL });
  } catch (e) {
    loader.fail(`we found an error: ${e}`);
  }

  return parameters.length ? content.join('\n') : '';
}

const isValidGroupBy = (groupBy: string | undefined): groupBy is keyof typeof GroupBy | undefined => {
  return groupBy === undefined || Object.keys(GroupBy).includes(groupBy);
};

interface CommandList extends Command {
  groupBy?: string
};

export const command = async (command: CommandList): Promise<void> => {

  if (!isValidGroupBy(command.groupBy)) {
    console.error(`groupBy is no valid, please try ${Object.keys(GroupBy).join(' | ')}`);
    process.exit(1);
  }

  const { params, credentials } = await getGlobalOptions(command);

  setAWSCredentials(credentials);

  const response = await listParameters({ ...params, groupBy: command.groupBy });

  console.log(response);
};