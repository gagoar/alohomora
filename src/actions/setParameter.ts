import SSM from 'aws-sdk/clients/ssm';
import dateFormat from 'dateformat'
import Table from 'cli-table3';
import ora from 'ora';

import { Actions } from '../types';
import { REGION, API_VERSION, Environment, DATE_FORMAT, SUCCESS_SYMBOL, DISABLE_TABLE_COLORS } from '../utils/constants';
import { getGlobalOptions, Command } from '../utils/getGlobalOptions';
import { setAWSCredentials } from '../utils/setAWSCredentials';


interface Input extends Actions {
  name: string;
  value: string;

  description?: string;
};

export const setParameter = async ({ name, value, description, prefix, region = REGION, ci = false, environment = Environment.all }: Input): Promise<string> => {

  const loader = ora(`storing key ${name} with the prefix /${prefix}  (${region})`).start();

  const ssm = new SSM({ apiVersion: API_VERSION, region });

  const keyName = `/${prefix}/${environment}/${name}`;

  const params = {
    Name: keyName,
    Value: value,
    DataType: 'text',
    Description: description,
    Overwrite: true,
    Type: 'SecureString'
  }

  const table = new Table({
    head: ['Name', 'Value', 'Environment', 'Updated at', 'Version'],
    style: ci ? DISABLE_TABLE_COLORS : undefined
  });

  let response: SSM.PutParameterResult | undefined;
  try {
    response = await ssm.putParameter(params).promise();

    table.push([name, value, environment, dateFormat(Date.now(), DATE_FORMAT), response.Version]);

    loader.stopAndPersist({ text: `stored ${name}, under /${prefix}  (${region})`, symbol: SUCCESS_SYMBOL });
  } catch (e) {
    loader.fail(`Something went wrong storing the key ${e}`);
  }

  return response ? table.toString() : '';
}


export const command = async (name: string, value: string, description: string | undefined, command: Command): Promise<void> => {

  const { params, credentials } = await getGlobalOptions(command);

  setAWSCredentials(credentials);
  const response = await setParameter({ ...params, name, value, description: description });
  console.log(response);
}