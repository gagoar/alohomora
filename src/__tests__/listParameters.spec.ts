import listParametersPayload from '../__mocks__/describeParameters.json';
import SSM from '../__mocks__/aws-sdk/clients/ssm';
import { listParameters } from '../';
import { mockProcessExit } from 'jest-mock-process';
import { listCommand } from '../actions/commands';
import { mockConsole, unMockConsole, createHandler } from './helpers';

const mockSSResponse = () => {
  SSM.__setResponseForMethods({ describeParameters: createHandler(() => ({ Parameters: listParametersPayload })) });
}
describe('listParameters', () => {
  let consoleErrorMock: jest.Mock;
  let consoleLogMock: jest.Mock;

  const prefix = 'my-company/my-app';
  beforeAll(() => {
    consoleLogMock = mockConsole('log');
    consoleErrorMock = mockConsole('error');
  });

  afterAll(() => {
    unMockConsole('log');
    unMockConsole('error');
  });

  beforeEach(() => {
    consoleLogMock.mockReset();
    consoleErrorMock.mockReset();
  });
  it('request fails', async () => {

    SSM.__setResponseForMethods({
      describeParameters: createHandler(() => {
        throw new Error('Some fatal error ocurred');
      })
    });

    const response = await listParameters({ prefix, ci: true });
    expect(response).toMatchSnapshot();
  });

  it('gets parameters', async () => {

    mockSSResponse()

    const response = await listParameters({
      prefix,
      environment: 'production',
      ci: true,
    });
    expect(response).toMatchSnapshot();
  });

  it('gets parameters, grouped by Name', async () => {

    mockSSResponse();
    const response = await listParameters({
      prefix,
      environment: 'production',
      groupBy: 'name',
      ci: true,
    });
    expect(response).toMatchSnapshot();
  });

  it('gets parameters, grouped by Environment', async () => {

    mockSSResponse()
    const response = await listParameters({
      prefix,
      environment: 'production',
      groupBy: 'environment',
      ci: true,
    });
    expect(response).toMatchSnapshot();
  });
  it('gets parameters, using nextToken', async () => {
    const handler = jest
      .fn()
      .mockImplementationOnce(() => ({
        Parameters: listParametersPayload.slice(0, 4),
        NextToken: 'MokeskinPouch',
      }))
      .mockImplementationOnce(() => ({
        Parameters: listParametersPayload.slice(5, 9),
      }));

    SSM.__setResponseForMethods({ describeParameters: handler });

    const response = await listParameters({ prefix, ci: true });
    expect(response).toMatchSnapshot();
    expect(handler).toHaveBeenCalledTimes(2);
  });

  it('via command invocation', async () => {
    mockSSResponse();

    await listCommand({ parent: { prefix } });

    expect(consoleLogMock).toHaveBeenCalled();
  });

  it('via command invocation, with an invalid groupBy set', async () => {
    const mockExit = mockProcessExit();
    mockSSResponse();
    await listCommand({ parent: { prefix }, groupBy: 'Bogart' });

    expect(mockExit).toHaveBeenCalledWith(1);
    expect(consoleErrorMock.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          "groupBy is no valid, please try name | environment",
        ],
      ]
    `);
  });

  it('via command invocation, with a valid groupBy', async () => {
    const mockExit = mockProcessExit();
    mockSSResponse()

    await listCommand({ parent: { prefix }, groupBy: 'name' });

    expect(mockExit).not.toHaveBeenCalled();

  });
});
