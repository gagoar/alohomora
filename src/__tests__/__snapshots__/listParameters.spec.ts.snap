// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`listParameters gets parameters 1`] = `
"[90m┌───────────────────────────────────────────────[39m[90m┬─────────────[39m[90m┬────────────[39m[90m┬──────────────────────────────────────────────────────┐[39m
[90m│[39m[31m Name                                          [39m[90m│[39m[31m Environment [39m[90m│[39m[31m Updated by [39m[90m│[39m[31m Updated at                                           [39m[90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m DEPLOY_SERVICE_STEP_FUNCTION_NAME             [90m│[39m development [90m│[39m severus    [90m│[39m Friday, July 12th, 2019, 12:21:37 AM (GMT+0000)      [90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m DEPLOY_SERVICE_STEP_FUNCTION_WRAPPER_NAME     [90m│[39m development [90m│[39m severus    [90m│[39m Thursday, October 3rd, 2019, 7:44:53 PM (GMT+0000)   [90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m GRAPHQL_ENDPOINT                              [90m│[39m all         [90m│[39m serious    [90m│[39m Thursday, November 29th, 2018, 9:26:47 PM (GMT+0000) [90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m GRAPHQL_ENDPOINT                              [90m│[39m production  [90m│[39m serious    [90m│[39m Thursday, November 29th, 2018, 9:26:47 PM (GMT+0000) [90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m PAGERDUTY_E2E_STEP_FUNCTION_INTEGRATION_TOKEN [90m│[39m all         [90m│[39m rubeus     [90m│[39m Monday, April 13th, 2020, 10:49:25 PM (GMT+0000)     [90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m PAGERDUTY_E2E_TEST_FAILURES_INTEGRATION_TOKEN [90m│[39m all         [90m│[39m rubeus     [90m│[39m Tuesday, April 21st, 2020, 9:32:08 PM (GMT+0000)     [90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m PHABRICATOR_CONDUIT_TOKEN                     [90m│[39m all         [90m│[39m serious    [90m│[39m Friday, November 23rd, 2018, 7:25:09 PM (GMT+0000)   [90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m PRISMA_ENDPOINT                               [90m│[39m all         [90m│[39m serious    [90m│[39m Friday, November 23rd, 2018, 7:26:02 PM (GMT+0000)   [90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m PRISMA_MANAGEMENT_API_SECRET                  [90m│[39m all         [90m│[39m serious    [90m│[39m Tuesday, January 22nd, 2019, 8:46:29 PM (GMT+0000)   [90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m SENTRY_DSN                                    [90m│[39m all         [90m│[39m serious    [90m│[39m Friday, November 23rd, 2018, 7:27:19 PM (GMT+0000)   [90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m SENTRY_ENVIRONMENT                            [90m│[39m all         [90m│[39m serious    [90m│[39m Sunday, November 25th, 2018, 9:11:45 PM (GMT+0000)   [90m│[39m
[90m└───────────────────────────────────────────────[39m[90m┴─────────────[39m[90m┴────────────[39m[90m┴──────────────────────────────────────────────────────┘[39m"
`;

exports[`listParameters gets parameters, using nextToken 1`] = `
"[90m┌───────────────────────────────────────────────[39m[90m┬─────────────[39m[90m┬────────────[39m[90m┬──────────────────────────────────────────────────────┐[39m
[90m│[39m[31m Name                                          [39m[90m│[39m[31m Environment [39m[90m│[39m[31m Updated by [39m[90m│[39m[31m Updated at                                           [39m[90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m DEPLOY_SERVICE_STEP_FUNCTION_NAME             [90m│[39m development [90m│[39m severus    [90m│[39m Friday, July 12th, 2019, 12:21:37 AM (GMT+0000)      [90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m GRAPHQL_ENDPOINT                              [90m│[39m all         [90m│[39m serious    [90m│[39m Thursday, November 29th, 2018, 9:26:47 PM (GMT+0000) [90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m PAGERDUTY_E2E_STEP_FUNCTION_INTEGRATION_TOKEN [90m│[39m all         [90m│[39m rubeus     [90m│[39m Monday, April 13th, 2020, 10:49:25 PM (GMT+0000)     [90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m PAGERDUTY_E2E_TEST_FAILURES_INTEGRATION_TOKEN [90m│[39m all         [90m│[39m rubeus     [90m│[39m Tuesday, April 21st, 2020, 9:32:08 PM (GMT+0000)     [90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m PHABRICATOR_CONDUIT_TOKEN                     [90m│[39m all         [90m│[39m serious    [90m│[39m Friday, November 23rd, 2018, 7:25:09 PM (GMT+0000)   [90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m PRISMA_MANAGEMENT_API_SECRET                  [90m│[39m all         [90m│[39m serious    [90m│[39m Tuesday, January 22nd, 2019, 8:46:29 PM (GMT+0000)   [90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m SENTRY_DSN                                    [90m│[39m all         [90m│[39m serious    [90m│[39m Friday, November 23rd, 2018, 7:27:19 PM (GMT+0000)   [90m│[39m
[90m├───────────────────────────────────────────────[39m[90m┼─────────────[39m[90m┼────────────[39m[90m┼──────────────────────────────────────────────────────┤[39m
[90m│[39m SENTRY_ENVIRONMENT                            [90m│[39m all         [90m│[39m serious    [90m│[39m Sunday, November 25th, 2018, 9:11:45 PM (GMT+0000)   [90m│[39m
[90m└───────────────────────────────────────────────[39m[90m┴─────────────[39m[90m┴────────────[39m[90m┴──────────────────────────────────────────────────────┘[39m"
`;

exports[`listParameters request fails 1`] = `""`;
